ompi/mca/common/ompio/common_ompio_request.c

```diff
@@ -125,33 +134,97 @@ void mca_common_ompio_request_alloc ( mca_ompio_request_t **req, mca_ompio_reque
 
 void mca_common_ompio_register_progress ( void ) 
 {
-    if ( false == mca_common_ompio_progress_is_registered) {
+    if (false == mca_common_ompio_progress_is_registered) {
+        OPAL_THREAD_LOCK (&mca_common_ompio_mutex);
+        if (mca_common_ompio_progress_is_registered) {
+            OPAL_THREAD_UNLOCK (&mca_common_ompio_mutex);
+            return;
+        }
         opal_progress_register (mca_common_ompio_progress);
```

Comment:
Is there a place where this callback is unregistered (I can't find it in this PR).

Reply:
that's also a good question, ompio so far has not unregistered the progress function again. My assumption was that an application is not just using  one non-blocking operations, but will most likely do that throughout the lifetime of the application, and hence we don't necessarily want to constantly register/unregister the progress function. I can add removing the progress function if the internal request list is empty. 

Are other components doing that (i.e. unregistering their progress function)?

Reply:
Most components unregister their progress function during shutdown. That way you avoid getting called after shutdown.

Removing progress callbacks seems quite expensive (linear search and O(n) atomic swaps) so doing that frequently is not a good idea.

Reply:
ok, thanks,  I can do the same. I will add the unregister functionality to the component shutdown function.

Reply:
We use the register/unregister of progress in many places. Take a look at the non-blocking communicator creation (in communicator/comm_request.c) for a sample.

Reply:
Can I suggest a two step plan here? 

- I've added the progress_unregister into the shutdown step of the component, and have done a fairly extensive round of testing with our internal testsuite, the ompio file I/O testsuite, and the hdf5 testsuite, and everything still seems to work. Since ompio so far has also not unregistered the progress function at runtime, this pr does not change the behavior of ompio (if anything, it improves it a bit by doing an unregister at all).

- After this pr, I can investigate the impact of moving the unregister operation to request finalize operation, i.e. if there is no pending I/o request anymore, we remove the progress function. I will look both at the impact on file I/O operations, but also on communication operations. We can than revisit the decision based on the data, which might point in either direction. (It might be however 1-2 weeks until I get to this task)

Would this be a reasonable path forward?

Reply:
Sounds good to me.